<font color=red><b>Delete the text marked <font color = blue>&lt;Your response here&gt;</font> and fill them in with your own responses. You are free to modify the format of this notebook to fit your needs, but must have responses for all of the questions somewhere in your notebook.</b></font>

# RSA Programming Project

Name:  <font color=blue>&lt;Theo Shin&gt;</font><br>
Moodle Email:  <font color=blue>&lt;tesh0094@colorado.edu&gt;</font><br>
Did you name your file "cuidentikey".ipynb ? Answer yes or no:  <font color=blue>&lt;yes&gt;</font>

---
# Part 1:

We will be creating our own RSA code to encode and decode secret messages. You are expected to write the code on your own as directed and not use direct sources from the internet. I do want you to use resources for understanding, but any copying and pasting, direct use of other code will be dealt with in the honor code office.  You may only use gcd() and % from math.  

RSA can seem complex, but we will be breaking it down into essential components or tools.

For this first part of the assignment you will write the following functions as directed.

Code must be in Python3 and use the algorithms and names directed. 
For each function, give a few comments at the top and comment each line.

You may use any IDE to develop but you must turn it in as a Jupyter notebook. You all have access to Jupyter notebooks at https://coding.csel.io

Follow the inputs and structure of the given algorithms.

IMPORTANT: It is essential to break this into small doable pieces. Do not try to write the whole thing at once. After the first tool kit is complete, the “meat” or RSA is the 3rd tool set.  Use examples from the book, notes, and homework to check that these pieces are working.  Then test that you can successfully code and decode simple numeric messages. Finally, add the text translation from the first tool kit.

---
## Introduction to the project and list of contents:

<font color=blue>&lt;To go step by step through the toolsets as we progress through the textbook and HW assignments. Being new to programming, I will definitely want to make sure I have the time to process and understand the correct implementation of this project. The most valuable takeaway for me is to get more comfortable with how to code and the many nuances of programming. I expect this RSA project to be very challening but am anticipating lots of insights to be gained. The table of contents will follow the default format: First Toolset (Convert_Text, Convert_Num, Convert_Binary String) Second Toolset (div_alg, FME, Euclidean_Alg) Third Toolset (Finding Private and Public Keys) Fourth Toolset (Encoding and Decoding Messages) and lastly wrapping it all up into one main function with all previous functions defined. I will doing everything on the jupyter hub as I don't have much familiarity with other IDEs.  Following the answers to project questions will be the factorize and factorize better functions and lastly, the conclusion. Here....we...go... &gt;</font>

---

### First Toolset
(You can do this now)

These are functions that you'll need to pre-process the messages before the messages are encoded and decoded by the RSA algorithm. Thats the reason we will be defining them first.

In [4]:
def Convert_Text(_string):
    """
    Define this function such that it takes in a simple string such as "hello" and 
    outputs the corresponding list of integers (ascii) for each letter in 
    the word hello. You may use ord(). Keep messages simple - only text and spaces.
    
    Args:
        _string (str): A message string

    Returns:
        list: A list of integers (ascii) corresponding to each letter in the input string
    
    Example:
        _string => hello
        returns => [104, 101, 108, 108, 111]
    """
    integer_list = [ord(a) for a in _string]  #ord() for each letter in string, returning an integer representing Unicode code.
                                              #https://www.programiz.com/python-programming/methods/built-in/ord
    
    return integer_list

print(Convert_Text("JUST DO IT"))

[74, 85, 83, 84, 32, 68, 79, 32, 73, 84]


In [5]:
def Convert_Num(_list):
    """
    Do the opposite of what you did in the Convert_Text function defined above.
    Define this function such that it takes in a simple string such as "hello" and outputs 
    the corresponding list of integers for each letter in the word hello.
 
    Args:
        _list (list): A list of ascii numbers

    Returns:
        string: A string of letters corresponding to the input list of ascii numbers
    
    Example:
        _list => [104, 101, 108, 108, 111]
        returns => "hello"
    """
    
    _string = ''.join(chr(a) for a in _list)    #concatenating chr() elements where unicode code represents a character 
                                                # https://www.programiz.com/python-programming/methods/built-in/chr
    
    return _string

print(Convert_Num([105]))

i


In [6]:
def Convert_Binary_String(n): # modified _int parameter to exponent "n" for congruency with FME function
    """
    Here, you need to define a function that converts an integer to a string of it's binary 
    expansion.
    
    Args:
        _int (int): An integer
    
    Returns:
        string: A string correspoding to the binary expansion of the input integer
    
    Example:
        _int => 345
        returns => "101011001"
    """

    
    string = ""
    while n > 0:           #iterative condition for while loop
        if n % 2 == 1:
            string += "1"
            n -= 1           #if number is odd, binary represenation will end in '1'
        else:
            string += "0"   #even binary numbers will end in '0'
        n = n/2             
    return string
 # [67, 111, 111, 108, 46, 32, 73, 116, 39, 115, 32, 111, 110, 32, 109, 121, 32, 108, 105, 115, 116]


print(Convert_Binary_String(111))

1111011


Now that you're done with the first toolset we'll move on to the second toolset which are actually involved in the RSA system.

---
### Second Toolset
(Note: the book has notes about its Pseudocode in the appendix)

In [7]:
def div_alg(c, d):
    """
    - This is a brute force algorithm that calculates c/d and remainder c%d.
    - The function must return two integers x and y such that x = c/d and y = c%d
      without the use of mod.
    - Use the algorithm 4 from page 253 of the book. I will allow the use of “%”
      in your code but I want you to code this as well. 
    
    Args:
        c (int): Dividend
        d (int): Divisor
    
    Returns:
        int, int: Two integers corresponding to c/d and c%d, without the use of mod
    """
#Algorithm 4: p.253
    '''x = 0
    y = abs(c)
    while y > d:
        y = y - d
        x = x + 1
    if c < 0 and y > 0:
        y = d - y
        x = -(x+1)
    return (x,y)'''    
    
    while d > 0:       #kept it simple. div and mod without divmod function. did not fully implement this function.
        x = c // d
        y = c % d
        break
    return x, y

print(div_alg(30,16))
print(div_alg(30,15))

(1, 14)
(2, 0)


In [2]:
def FME( b, n, m):
    """
    - Using the fast modular exponentiation algorithm, this function should return
      b**n mod m. Use the algorithm 5 on page 254 of the book (however, you may
      input the exponent and then convert it to a string - the book version imports 
      the binary expansion)  
    - You should use the function defined above: Convert_Binary_String()
    - Use this string to test which components are used in the calculation.
    - Yes, there are many other ways to do this. I am asking you to do it this
      unusual way on purpose.
    
    Args:
        b (int): Base
        n (int): Exponent
        m (int): Modulus
    
    Returns:
        int: The result of b**n mod m
    """ 
def Convert_Binary_String(n):      #implementing binary expansion function from (3)
    
    string = ""                   #empty string to characters based on binary conditions
    while n > 0:                  # modified _int parameter to exponent "n" for congruency with FME function
        if n % 2 == 1:            #if odd
            string += "1"         #include "1" to string
            n -= 1
        else:
            string += "0"         #if even, append "0"
        n = n/2
    return string                 #returns string value


def FME( b, n, m):
    str = Convert_Binary_String(n)  # calling binary conversion function from above
      
    int = 1                         # Use this to store the current product of terms
    for i in range(len(str)-1):     # starting from right-most side for binary represenations
        if str[i] == 1:             #if binary is "1"
             n += (2 ** (str[i]))   #exponent is equal to 2 ** nth position from right to left
        int = b ** n % m 
        
    return int                      #return b^n mod m value
print(FME(3,11,5))

2


In [2]:
def Euclidean_Alg(a, b):
    """
    - Calculate the Greatest Common Divisor of a and b.
    - Unless b==0, the result will have the same sign as b (so that when b is divided 
      by it, the result comes out positive).
    - The function must return a single integer 'x' which is the gcd of a and b.
    - Implement Algorithm 1 as Euclidean_Alg from page 269. Ultimately you will input 
      the binary expansion as a string (see 6) so keep that in mind. You can practice 
      it however you like though.)
    
    Args:
        a (int): An integer
        b (int): An integer
    
    Returns:
        int: Greatest Common Divisor of a and b
    """
    # SIMPLIFIED VERSION: 
    #if b == 0:
    #    return a
    #else:
    #   return Euclidean_Alg(b,a%b)
    
    if a > b:           #stating the lower of a,b
        result = b
    result = a

    if result == 1:     #if a,b is equal to 1, return 1
        return 1

    while result > 0:   #while a,b are positive
        if a % result == 0 and b % result == 0:     #if a mod b= 0 and b mod a =0
            return result
        result = result - 1
    
print(Euclidean_Alg(12,17))
        

1


---
### Third Toolset
(MUST USE your results from previous work)

Here we will implement the meat of the RSA cryptosystem. The functions below will generate the public and private key pairs which will then be used to create a ciphertext using the public key and then decode the same using the private key. 

In [10]:
def Find_Public_Key_e(p, q):
    """
    - Write a function Find_Public_Key_e that takes 2 prime integers p and q and returns 
      integer Public_Key_e. What is commonly called e.
    - Use the gcd function that you have defined before.
    - The function should return the public keys n and e
    - HINT: this function will run a loop to find e such that e is relatively prime to 
      (p - 1) (q - 1) and not equal to p or q.
      
    Args:
        p (int): Prime integer
        q (int): Prime integer
    
    Returns:
        int, int: The public keys n and e
    """
        
import random                               #import random module to derive random e

def Euclidean_Alg(a,b):                     #using gcd function
    if a > b:                               #stating the lower of a,b
        result = b
    result = a

    if result == 1:                         #if a,b is equal to 1, return 1
        return 1

    while result > 0:                       #while a,b are positive
        if a % result == 0 and b % result == 0:     #if a mod b= 0 and b mod a =0
            return result
        result = result - 1

def Find_Public_Key_e(p,q):                    
    n = p * q                              #assigning n variable as p times q
    phi = (p-1) * (q-1)                 #assigning phi variable
    e = phi
    while Euclidean_Alg(phi,e) != 1:      # 1<e<phi  ensuring e and phi do not have any common divsors except 1
        e = random.randint(3,phi-1)   
    return(n,e)
    
print(Find_Public_Key_e(61,89))

(5429, 3431)


In [8]:
def Find_Private_Key_d(e,p,q)

#e must be relatively prime to phi(n)
#Referenced and used https://brilliant.org/wiki/extended-euclidean-algorithm/ for extended gcd pseudocode; I ultimately tried to involve my own e_gcd algorithm,
#but had already spent a great deal of time trying to write this Python code. I was essentially stuck on deriving an algorithm and needed to progress to have a 
#chance at completing the project by its due date. I understand the algorithm as a recursive loop to determine the Bezout coefficients
#greatest common divisor D = ax + by. From HW 7, working with Theorem 2 we also know the prime divisor is 1<a<n and <= sqrt(n).

def egcd(a, b):                            #derived egcd function       
    x,y, u,v = 0,1, 1,0                    #assiging x = 0, y =1, u=1, v=0
    while a != 0:
        q, r = b//a, b%a                   #assigning q and r values and div and mod 
        m, n = x-u*q, y-v*q                 
        b,a, x,y, u,v = a,r, u,v, m,n        
    gcd = b
    return gcd, x, y                      #return values to incorporate into Find_Private_Key_d function 



def Find_Private_Key_d(e, p, q):
    phi_n = (p-1) * (q-1)           #taking p and q values to derive phi
    gcd, x, y = egcd(e, phi_n)      #using  extended gcd function for e and phi
    if gcd != 1:abs                 #ensuring relatively prime
        raise Exception('modular inverse does not exist')          #if not relatively prime
    else:
        return x % phi_n    #e,#phi 
print(Find_Private_Key_d(169, 13, 23))  
 #   d= e^−1 mod ϕ(n)  

---
### Fourth Toolset
Must use results from Find_Public_Key_e <br>
(to test - just use numbers at first before including the text translations)

1. In this part, you will define two functions `Encode` and `Decode` which will use the public and private keys that you calculated using the above 2 functions in the third toolset.
2. Using the public key, the `Encode` function will encode a message and generate the corresponding cipher_text.
3. Using the private key, the `Decode` function will decode a ciper_text and recover the original message.


In [9]:
def Encode(n, e, message):
    """
    - This function takes a string and the public key and returns a coded message as
      a list using Convert_Text, Convert_Binary_String, and FME.
    - Use the function Convert_Text to get a list of numbers.
    - Encode each of those numbers using n and e and return the encoded cipher_text.
    - Use Convert_Binary_String to convert the exponent to a string and use that as 
      part of FME (yes, there are many other ways to do this, but I am asking you to
      use a string).
    - For this implementation you will be taking a string, converting it to a list of 
      ascii numbers, then “RSA-ing” each number(letter)on the list, then returning 
      the encoded list.
    
    Args:
        n (int): Public key modulus
        e (int): Public key exponent
        message (str): A message string to be encoded
    
    Returns:
        list: A list containing the encoded cipher text
    """
    
#From here on out, I could not get my python codes to use funciton results as values for other functions. I also struggle with this in my CSCI 1300
#class which I plan on revisiting these functions once I gain a thorough understanding of nested functions or just correctly using the function results. The manner in 
#which I encoded messages for the piazza discussions was by using provided public/private keys, using Convert_Text function, then implementing the FME function. Thus,
# I was not able to get the function to provide a list of encoded cipher text and instead manually inputted the numbers derived from Convert_Text. The layout and 
# understanding of how encode function operates is better defined below for the "main" function but I wanted to get the basic layout down here for reference.
    
import random                                      #import random module

def Convert_Text(message):                         #expressing Convert_Text message for toolset 1

    
    integer_list = [ord(a) for a in message]  
    
    return integer_list



def Convert_Binary_String(integer_list):  #Taking integer_list and converting to binary equivalent
    
    string = ""
    while n > 0:     # modified _int parameter to exponent "n" for congruency with FME function
        if n % 2 == 1:
            string += "1"
            n -= 1
        else:
            string += "0"
        n = n/2
    return string     # used for FME function below  

    
def Euclidean_Alg(a,b):               #Defining Euclidean_Alg function from toolset 2 to ensure prime factors
    if b==0:
        return a
    else:
        return Euclidean_Alg(b,a%b)

def Find_Public_Key_e(p,q):              #getting public keys for fast modular exponentiation for final step in encoding message
    n = p * q   
    phi = (p-1) * (q-1)
    e = phi
    while Euclidean_Alg(phi,e) != 1:
        e = random.randint(3,phi-1)   #1<e<phi
    return(n,e)
    



def FME( b, n, m):
    #m^e mod n = c
    bin_string = Convert_Binary_String(n)               #from Convert_Binary_String following text conversion
    e = Find_Public_Key_e    # 2nd return value from above function
    n = Find_Public_Key_e    # 1st return value from above function
    
    message = 1  # storing initial int value
    for i in range(len(bin_string)-1):     #stating index from right-hand side
        if bin_string[i] == 1:     #if index value contains 1
            e += (2 ** (bin_string[i]))      #exponent is equal to 2 ** nth position from right to left
        message = bin_string ** e % n     
        
    return message


In [10]:
def Decode(n, d, cipher_text):
    """
    - This function takes the list of coded numbers, both private and public keys, and 
      returns decrypted message as a string (use Convert_Num and FME).
    - Here, the cipher_text will be a list of integers.
    - First, you will decrypt each of those integers using n and d.
    - Later, you will need to use the function that converts the integers to a string 
      defined in second toolset to recover the original message as a string. 
    
    Args:
        n (int): Public key modulus
        d (int): Private key
        cipher_text (list): A list containing the encoded cipher text
    
    Returns:
        str: A string containing a decrypted message
    """
    
#As descrived in the "break codes" explanation below, in order to "crack" Amanda and Isaac's messages, I used the brute force factorization to derive p, then divide n by p to get q. Using the Find_Private_Key_d function, I was
#then able to derive phi and get the resulting private key d. Next, use cypher text, private key, and provided n value from the public key to get ascii numbers. And 
#finally converting the numbers with Convert_Num function. Again, proper layout and more coherent understanding of operation is in "main" function but I wanted 
# to get the general gist of functions used here.


    
def egcd(a, b):                          #Incorporating extended GCD to derive private key.                           
    x,y, u,v = 0,1, 1,0
    while a != 0:
        q, r = b//a, b%a
        m, n = x-u*q, y-v*q
        b,a, x,y, u,v = a,r, u,v, m,n
    gcd = b
    return gcd, x, y



def Find_Private_Key_d(e, p, q):
    phi_n = (p-1) * (q-1)
    gcd, x, y = egcd(e, phi_n)
    if gcd != 1:
        raise Exception('modular inverse does not exist')          #if not relatively prime
    else:
        return x % phi_n    #e,#phi              #once private key is established, plug values into FME function and convert using Convert_Num function

    def FME( b, n, m):
        b = int(input("enter cypher text"))                        #prompt user for cypher text message (used piazza posts for testing)
        n = int(input("enter provided private key d:"))            #prompt user for value d derived from Find_Private_Key_d function above
        m = int(input("enter provided n:"))                        #prompt user for provided n value
    
        str = Convert_Binary_String(n)  # calling binary conversion function from above
      
        asc = 1  # Use this to store the current product of terms
        for i in range(len(str)-1):
            if str[i] == 1:
                 n += (2 ** (str[i])) 
            asc = b ** n % m 
        
        return asc
        print(asc)

    def Convert_Num(_list):               #Convert derived ascii integers into string of letters/words

       # _list = FME(asc)   #would normally call within main function
    
        _string = ''.join(chr(a) for a in _list)
    
        return _string
    
print (FME(85226, 115979, 639629))
print(Convert_Num([83]))

    return message

---
### Now the fun begins:

In [11]:
if __name__ == '__main':
    """
    Implement a main function such that:
    
    1. Asks the user to Get Keys, Encode or Decode.
    2. If Getting keys, it will need to ask for p and q. 
    3. If Encoding, it will need the message and public keys.
    3. If Decoding it will need the coded message, and public and private keys.
    
    You may implement this differently if you want, add more functions, 
    include a list of useful primes etc... 
    Just implement in a way that makes sense to you.
    And explain why.
    
    ALSO - IF YOU ARE NEW TO PYTHON. 
    If you are struggling to get the main function working, go ahead and use the functions you
    have created in the notebook to solve the RSA codes without a main function.
    """
#RSA Main Function

import random         #import random module
import math           #import math module



#Listing of functions from each toolset 1:

def Convert_Text(_string):
#     """returns list of corresponding ascii integers for each letter in word
#     Parameters(1): string 
#     Returns: Corresponding list of ascii integers"""

    integer_list = [ord(a) for a in _string]  
    
    return integer_list


def Convert_Num(_list):
#     """returns corresponding letter for each ascii integer
#     Parameters(1): list of ascii numbers 
#     Returns: String of letters corresponding to input of ascii numbers"""
    
    _string = ''.join(chr(a) for a in _list)    
    
    return _string
                                                  


def Convert_Binary_String(n): # modified _int parameter to exponent "n" for congruency with FME function
#     """returns string corresponding to binary expansion of input integer
#     Parameters(1): An integer
#     Returns: Binary expansion"""   


    
    string = ""
    while n > 0:           #iterative condition for while loop
        if n % 2 == 1:
            string += "1"
            n -= 1           #if number is odd, binary represenation will end in '1'
        else:
            string += "0"   #even binary numbers will end in '0'
        n = n/2             
    return string


#Listing of functions from each toolset 2:

def div_alg(c, d):
#     """returns two integers corresponding to c/d and c%d
#     Parameters(1): dividend(c) and divisor(d)
#     Returns: div and mod values"""   

      
    while d > 0:      
        x = c // d
        y = c % d
        break
    return x, y    

def FME( b, n, m):
#     """returns result of b**n mod m
#     Parameters(1): base, exponent, modulus
#     Returns: result for fast modular exponentiation"""   


    str = string                   # calling binary conversion return value
      
    int = 1                         # Use this to store the current product of terms
    for i in range(len(str)-1):     # starting from right-most side for binary represenations
        if str[i] == 1:             #if binary is "1"
             n += (2 ** (str[i]))   #exponent is equal to 2 ** nth position from right to left
        int = b ** n % m 
        
    return int                      #return b^n mod m value

    
def Euclidean_Alg(a, b):
#     """returns greatest common divisor of a and b
#     Parameters(1): a and b integers
#     Returns: GCD of a and b"""

    
    if a > b:           #stating the lower of a,b
        result = b
    result = a

    if result == 1:     #if a,b is equal to 1, return 1
        return 1

    while result > 0:   #while a,b are positive
        if a % result == 0 and b % result == 0:     #if a mod b= 0 and b mod a =0
            return result
        result = result - 1    

        
#Listing of functions from each toolset 3 (along with factorize_n):

def Find_Public_Key_e(p, q):
#     """takes p and q and returns public keys n and e
#     Parameters(1): p and q (prime)
#     Returns: public keys"""

    n = p * q   
    phi = (p-1) * (q-1)
    e = phi
    while Euclidean_Alg(phi,e) != 1:      # 1<e<phi  ensuring e and phi do not have any common divsors except 1
        e = random.randint(3,phi-1)   
    return(n,e)
        

    

def egcd(a, b):    
#     """calculates greatest common divisor from extended GCD
#     Parameters(1): integers a and b
#     Returns: GCD value"""
    x,y, u,v = 0,1, 1,0
    while a != 0:
        q, r = b//a, b%a
        m, n = x-u*q, y-v*q
        b,a, x,y, u,v = a,r, u,v, m,n
    gcd = b
    return gcd, x, y



def Find_Private_Key_d(e, p, q):
#     """calculates private key d using extended GCD
#     Parameters(1): exponent e, p, and q
#     Returns: d - modular inverse of e"""

    phi_n = (p-1) * (q-1)
    gcd, x, y = egcd(e, phi_n)
    if gcd != 1:
        raise Exception('modular inverse does not exist')          #if not relatively prime
    else:
        return x % phi_n    #e,#phi 

def factorize(n):
#     """returns prime factors of n
#     Parameters(1): n value
#     Returns: prime factors"""

    a = int(math.sqrt(n))
    if n % 2 == 0:
        return 2
    i = 3
    while i<=a:
        if n % i == 0:
            return i
        i += 2
    
    
#encode/decode/incorporating defined functions    
    
def Main():
#     """Main() function to store variables and call to functions
#     Parameters(0): none; Stores variables, call to functions, and prompts user whether to get keys, encode, or decode
#     Returns: private/public keys, encodes message, decodes message"""   
    
    print("Welcome to RSA \n")      #Title print statement followed by a new line
    
    prompt = input("get (k)eys, (e)ncode, or (d)ecode?:")              #prompt user for 3 options
    
    if prompt == k:                                   #if user inputted wanting to get keys:
        p == input("Value for p:")                    #prompt user for p value
        q == input("Value for q:")                    #prompt user for q value
        n = p * q                                     #assign variable n as p times q
        phi = (p-1) * (q-1)                           #assign phi as (p-1)(q-1)
        e = phi                                       
        while Euclidean_Alg(phi,e) != 1:              #Find_Public_key algorithm
            e = random.randint(3,phi-1)   
        return(n,e)                                   #Public Key values n and e
        if egcd(e,phi_n):
           private = Find_Private_Key_d(e,p,q)         #getting private key value
        print(n,e,private)                             #print result for public and private keys after user entered p and q
       
    if prompt == e:                                           #if user inputs encode
        m = input("Enter message for encoding:")              #prompts user for string needing encoding
        encoded_m = Convert_Text(m)                           #assigns variable encoded_m for ascii values
        prompt=input("Do you know the public keys (y/n)?:")   #assume user does not have public keys available then find public key using p and q
        if prompt == y:
            public_n = int(input("Enter provided n:"))
            public_e = int(input("Enter provided e:"))
        else:
            p_value = int(input("Enter value for p:"))
            q_value = int(input("Enter value for q:"))
            public_key = Find_Public_Key_e(p_value, q_value)
        print(FME(m,(reversed(public_key))))                    #to switch n,e to e,n ; prints result for encoded message

    if prompt == d:                                           #if user inputs decode
        m2 = input("Enter message for decoding:")             #asks user to input cypher text
        n2 = int(input("Enter known n value:"))               #asks user for n value
        e2 = int(input("Enter known e value:"))               #asks user for e value
        p2 = factorize(n)                                     #uses factorize brute force function 
        q2 = n//p                                             #calculate and assign q variable
        phi2 = (p2-1) * (q2-1)                                #calculate phi from q and factorized n result
        if egcd(e2,phi2) == 1:                                #if relatively prime
            private_key = Find_Private_Key_d(e2,p2,q2)        #determine private key from calculated e,p,q values
            asc_num_list = (FME(m2,private_key,n2))           #resulting list after Fast modular exponentiation
        print(Convert_Num(asc_num_list))                      #print converted ascii integer list to string values


        
    if __name__ == "__main__":  #using main function from inside statement(still determining how to incorporate the "main" function usage)
        main()
        
        

Demonstate how your RSA works below, by encoding and decoding a sample message here using your main above and sample inputs. Describe how this works and why you choose to put it together the way you did.

Notice you can run your code, organize how you did a specific examples etc, all in this notebook. So I can see how you put everything together - mixing code blocks and text blocks. You might find it helpful to use some of homework 7 for this - the part where you explain RSA. I am not using an autograder, so this demo should demonstrate with an example how your code works. Do an example in the notebook, leaving the inputs and results that I can retest as I review.

<font color=blue>&lt;Well...I gave a valiant effort, but was unable to get the "main" function to cooperate. I am new to python (2 months) and am hoping I can get a grasp of the main function before semester's end. I should be getting back my graded assignments using main for my CSCI 1300 class and will continue to revise this code for my own knowledge. I organized the RSA code by listing out all the functions from part 1 in pretty much the same order. Once I got to main, I intended to provide user with three options through a user prompt for getting keys, encoding, or decoding. Depending on their response, the program was intended to go through the necessary functions, get the need results, and print a statement with public keys if user chose "k", encoded message if "e", and decipher code if "d" chosen. Additional comments above in code.
    
Sample Message for Encoding: 
if prompt == e:
    m = "Curiosity killed the cat"
    encoded_m = Convert_Text(m)                                # converts "curiosity killed the cat" into list of ascii numbers  
        prompt=input("Do you know the public keys (y/n)?:")    # [67, 117, 114, 105, 111, 115, 105, 116, 121, 32, 75, 105, 108, 108.....]
        if prompt == y:
            public_n = int(input("Enter provided n:"))
            public_e = int(input("Enter provided e:"))
        else:
            p_value = int(input("Enter value for p:"))
            q_value = int(input("Enter value for q:"))
            public_key = Find_Public_Key_e(p_value, q_value)     # determines needed public key either from prompt or Find_Public_Key_e function
        print(FME(m,(reversed(public_key))))           # Returns cypher text from original message         
                                                        # [15908, 27872, 28128, 21476, 18423, 14591, 21476, 3532, 22024, 234.....]
    
Sample Message for Decoding:
     if prompt == d:
        m2 = input("Enter message for decoding:")          #Enter cypher text for decoding process
        n2 = int(input("Enter known n value:"))          #[13842, 9054, 23468, 5090, 27872, 28128, 21476, 18423, 27872, 14591, 23468....]
        e2 = int(input("Enter known e value:"))
        p2 = factorize(n)                                  #Brute force algorithm to determine factors to determine p and q
        q2 = n//p 
        phi2 = (p2-1) * (q2-1)
        if egcd(e2,phi2) == 1:
            private_key = Find_Private_Key_d(e2,p2,q2)    
            asc_num_list = (FME(m2,private_key,n2))       #FME print result: [66, 101, 32, 99, 117, 114, 105, 111, 117, 115, 32, 110, 111...]
        print(Convert_Num(asc_num_list))                  #Convert_Num print result: Be Curious not afraid
    #to switch&gt;</font><br>

---
## Part 2:



After you have tested your RSA package yourself, test it with classmates by publishing both the private and public keys on Piazza category for this (this will be a folder separate from Project) for a message. Then post 3 messages with just the public key for people to break - use both very small n ‘s (under 1000) for practice, and some with a challenge.

To break others’ code you just need to factor n (public key).

Write simple “brute force” method for finding for prime factors of n and use this to try to crack codes.
To improve this factoring algorithm, consider THM 2 from page 258.

In [1]:
import math
def factorize(n):
    """
    - Implement the basic brute force factorization algorithm shown as pseudo code
      below to find prime factors of n.
    
    Pseudocode:
        # n is a number, return the smallest factor of n
        for i from 2 to n-1:
            if i divides n:
                return i
        return FALSE"""

    
# from Theorem 2, if n is a composite integer, n has a prime divisor less than or equal to sqrt(n)
   #   result = 0

    
    a = int(math.sqrt(n))           
    if n % 2 == 0:            #if n even return 2
        return 2
    i = 3                  # counter
    while i<=a:              
        if n % i == 0:      #while loop to determin prime factors  
            return i  
        i += 2               
    
print(factorize(979))
print(299//13)


11
23


Next we will put it all together:

1.	Find your public and private keys with 6 and 7
2.	Create digital messages with 4 and encode with 8.
3.	Publish your public and key and message to Piazza.
4.	Exchange private key with other classmates to Piazza.
5.	Test!

Write a page or two long equivalent narrative of the results of exchanging keys with classmates and testing your code. What was most challenging? What was most satisfying? Who helped you? Which resources were most helpful?

<font color=blue>&lt;Phew!! This RSA project was no joke (not that I thought it would be). Though my RSA algorithms didn't churn out perfect results, I'm quite pleased with my progress in understanding not only RSA but python coding (This assignment was definitely harder than any of the CSCI 1300 HW assignments thus far). On Friday (3/15), even though my code wasn't up and running as smoothly as the finished product, I was able to start cracking the codes with provided private keys. I chose test my code with Samuel Cramer's crypted message. From his previous piazza messages, he seems to be a particularly bright student with a good sense of humor and I was actually interested in what type of message he would encrypt. I'll have to admit, the first time I went through the crytic messages, I was lost and completely bewildered lol. Maybe it was because it was a Friday but it took me at least 30 minutes before I was finally to understand the path from message to getting the ascii integer list and then converting the list into letters/words. I felt tremendous relief when the cyphered text actually returned coherent words and not just shapes and symbols which I routinely got in the beginning (due to using incorrect inputs for the FME algorithm). I was never able to get my functions to take an entire list of cyphered numbers and return a sentence. Instead, I resorted to manually going through the each number followed by commas and going through the steps one by one. Again, I think I'll continue to refer to this assignment in Jupyter and as my coding/knowledge progresses, I want to be able to configure the code how I intended where I can copy/paste a list of coded messages and have the algorithm return a simple message print statement. Samuel's eventual message read "Sirens of Titan is my favorite book, I recommend it."  I haven't actually read the book yet, so I get to working on my coded response. I used the Convert_Text function to get the ascii integers associated with the text "Cool. I'll add to my list." From there, I went through the process of encrypting the message. I wish I either had more time or a better understanding the implementation of these functions since I would have definitely enjoyed a more engaging exchange. However, I knew if I was limited by going through each step manually, I would be using up previous time to be used to ensure I complete the project to the best of my ability. Following my response, I made sure to post my public messages without private keys for other students to break. As of my writeup, I have not received a response. Again, my code has improved significanty since Friday so I'm sure I could now write more engaging comments/responses. My messages were related to sports predictions (NFL - BROWNS Super Bowls Champs for next year and the use of the Nike slogan). I continued to look for piazza posts with public and private keys to generate some confidence and understanding of deciphering messages. I chose to take a look at the instructor's message which generated ascii values [34, 73, 116, 32, 108, 111, 118, 101, 100, 32, 116, 111, 32, 104, 97, 112, 112, 101, 110, 46, 34] and the string "It loved to happen," a famous Marcus Aurelius quote. My response, from the movie Gladiator with the character Marcus Aurelius, stated "I will have my vengeance. In this life or the next." A chose another message from the instructor with the deciphered message: "Be Curious Not Afraid." At this point, I started to get a pretty grasp on going back and forth between encoding and decoding and encoded my response "Curiosity Killed the cat."  Most challenging for me was the Find_private_key_d function which I admittedly needed help with and also working on wrapping up all the functions into one coherent "main" function. Being new to coding, python, everything...I found every step along the way satisfying. The challenge obviously grew as the toolsets progressed and I found myself working harder and longer through each successive one. The most satisfying aspect was decoding the messages. I was worried I wouldn't get to that point and not be able to post anything for contribution. No particular student helped me but it was definitely helpful working along the homework assignments and incrementally gaining understanding of RSA. Aside from lectures, I referenced https://iinurmi.wordpress.com/2015/12/04/rsa-encryption-private-and-public-key-calculation/ continuously to keep familiarizing myself with RSA as the entire concept was not intuitive for me. I never knew a school assignment could be so challenging, frustrating, surprising, fun all in one. It was a great learning experience/tool and I intend to continue modifying this code as my skills progress&gt;</font><br>

Describe your attempts to “break” codes for which you do not have the private key. How large do numbers need to be before you could no longer find factors?

<font color=blue>&lt;I started with Amanda Killeen's message. I honestly did not think I stood a chance at breaking the code without the private key. That algorithm caused me a lot of pain/grief and time and as annotated in my comments for that finding private key d, I resorted to online tutorials for understanding how to implement the extended Euclidean algorithm for getting a defined private key. So, at first glance, even the her message was relatively short (around 16 characters), it took a while before I got the ball rolling. Obviously I had to define my private key function so I had something to work with. I first used the factorize_n function to determine p and q values. I was eventually able to implement a modular inverse type function. The first four tries or so, I kept on getting ascii integers which were way too many digits and resulting in text conversions that didn't make sense. So this took a lot of trial and error and seeing why I was getting the wrong values for d. Eventually I ended on a working private key function using values e and (p-1)*(q-1) I was able to get the ascii values [89, 111, 117, 39, 114, 101, 32, 109, 121, 32, 98, 111, 121, 32, 66, 108, 117, 101, 33] which I then converted to "You're my boy Blue!" Not gonna lie, I was pretty ecstatic when I saw the converted message as Old School is a cult classic. I posted in a piazza response the song played during the funeral scene where Will Ferrel's character, Frank the Tank, agonizes and says the famous line. My next attempt was from student Isaac Haames. Isaac's n and e values were smaller than Amanda's message and so I was hoping this one would require less effort as I knew I had to really work through the rest of the RSA project and the get the pieces working and finalized. With n = 299 and e = 169 I was able to factorize n = 13 and derive q = 23/ With these values I took the same process as before and got the deciphered message "Isn't RSA fun?" As for finding out how large numbers need to be before I received error "int too large to find float", I got to 310 characters.&gt;</font><br>

How many codes did you break? Give some examples.

<font color=blue>&lt;The two that I mentioned and if time allows or even after project submission, I'll use some of the student's posts for practice in code breaking when I improve my algorithm. Examples of codes broken from piazza posts: Amanda Killeen, public key e=13 n=5429. Calculated d using my algorithm to get d = 2437 and using brute force factorization, p=61 q=89 and phi=5280. The cipher text list was  [801, 50, 83, 5170, 1244, 2480, 4851, 1085, 4940, 4851, 4979, 50, 4940, 4851, 3211, 4866, 83, 2480, 2249] which translated to [89, 111, 117, 39, 114, 101, 32, 109, 121, 32, 98, 111, 121, 32, 66, 108, 117, 101, 33], "You're my boy Blue!"  My second code was from Isaac Haames, public key e = 169, n = 299. Calculated d as 25 and factorized n as p =13, q=23, and phi = 264. The cipher text was [225, 109, 273, 6, 248, 115, 6, 115, 20, 6, 291, 39, 60, 13, 6, 258, 39, 188] which I then got the ascii integers as [82, 83, 65, 32, 105, 115, 32, 115,111, 32, 109,117, 99,104, 32,102,117,110] translating to "RSA is so much fun"    &gt;</font><br>

Use THM 2 from page 258 to improve your brute force method by improving our code above. Explain with examples (doing an example in the notebook like #4) how using THM 2 page 258 can improve the brute force method breaking RSA code

In [24]:
import math
def factorizebetter(n):
# from Theorem 2, if n is a composite integer, n has a prime divisor less than or equal to sqrt(n)
#reference https://www.geeksforgeeks.org/ for understanding and providing factors list and running efficient algorithm for prime factors .


    i = 2                    
    factors = []                 #perform divisions of n by successive primes beginning with 2
    while i**2 <= n:          # from Theorem 2, if n is a composite integer, n has a prime divisor less than or equal to sqrt(n)
        if n % i:
            i += 1         #increment if remainder
        else:
            n = n // i       # continue dividing by successive primes ; append to list
            factors.append(i)
    if n > 1:
        factors.append(n)
    return factors
print(factorizebetter(7007))

[7, 7, 11, 13]


<font color=blue>&lt;THM 2 can improve brute force method for breaking RSA code similarly to what Hira speaks about the piazza RSA project discussion. Computers are better able to handle a lot of computations on smaller fixed sized chunks. My understanding is because THM 2 breaks down the factorization futher there's less computation work for larger n values.&gt;</font><br>

So at this point, just go as far you are interested in going. 

 

There is a whole world of research around looking for and factoring large primes because if you can do that you will break the global banking system.

 

Some things to think about when "really" trying to break RSA

 

(***VERY TOTALLY OPTIONAL - AVERT YOUR EYES***)

 

To improve the brute force method - a very naive and slow approach in many ways. We recommend considering improvements in the following order.

Basic Improvement: Skip over certain numbers:
The brute force approach above tests even numbers as well as odd. If the number is odd, there is no point searching for an even factor.
We can do a better job tackling multiples of 3. If the number is not a multiple of 3, then no factor can be a multiple of 3 either.
How about multiples of 5,7,11,13,…
Basic Improvement: The brute force approach iterates all the way to n−1n−1. Can we stop earlier?

 Advanced Ideas:

What other factoring algorithms are at our disposal? Do they work better.

There are a few that are fairly accessible. These include Pollard's-Rho algorithm. I have posted a tutorial on it here: http://www.cs.colorado.edu/~srirams/courses/csci2824-spr14/pollardsRho.html. My code is in the C language, but you will have to implement in Python3. So you should focus on the main ideas. There are other much better online expositions. Can you understand and implement it?

There are other approaches such as Fermat factoring, the number field sieve. But these are quite advanced.

The algorithm works on a single thread. Can you explore parallelizing this algorithm? You can read about how to write parallel multithreaded code in Python3 : https://www.tutorialspoint.com/python3/python_multithreading.htm.
Warning: This is really advanced stuff for beginner programmers, but perhaps never too early to learn some ideas on parallelization. Post on piazza for some ideas on how to parallelize brute force search. The tricky part is to stop all threads if one thread found a factor, look up threading.Event objects.

---
## Conclusion:

Return to the questions and comments you asked about RSA in homework 6.  Answer your questions and/or explain what you have learned.

<font color=blue>&lt;My observations on RSA from the HW assignment talked about my amazement towards RSA, especially deriving the private key and comparing the keys to complementary light colors. My views on the complexity of private key has only been illuminated from this project since I found deriving the private key value to be a tremendous challenge while using the extended Euclidean algorithm. I also mentioned in the homework how as the value of the keys grow, the time constraint to solve for the values grows. This was seen in the brute force factorizing of n where it takes longer to derive each value it spans many integers. Going from writing the homework assignments by hand to watching the lectures and not understanding what was happening at times, I felt this project was an enormous undertaking but feel RSA is something I can continue to delve into and learn to understand better. This provide provided a definite framework and template for breaking down RSA into manageable sections and building it back up into what makes sense for me given my limited experience. As stated, I would definitely like to improve these algorithms later down the road to derive more efficient codes and altogether better understanding of RSA. The main function still gives me issues and I definitely want to revisit the main code for when I get a better grasp on functions. &gt;</font><br>

---
## Grading

1. (2) Your name and Moodle email
2. (5) A short introduction (words and overall goals) and table of contents
3. (30) Your RSA package and all functions requested in RSA project part 1- 
4. (22) A short demo in a Jupyter block walking thru the steps of RSA and showing your code works. Notice you can run your code, organize how you did a specific examples etc, all in the notebook. So I can see how you put everything together - mixing code blocks and text blocks. You might find it helpful to  use some of homework 7 for this - the part where you explain RSA.

I am not using an autograder, so this demo in #4 should demonstrate with an example how your code works. DO an example in  the notebook, leaving the inputs and results and I can retest as I review.

5. (10) A two page long narrative of the results of exchanging keys with classmates and testing your code. What was most challenging?  What was most satisfying? Who helped you? Which resources were most helpful? 

**6 - 9 must be started before Friday midnight before the project is due**
6. (5) Include your simple code for finding factors of n. 
7. (5) Describe your attempts to “break” codes for which you do not have the private key. How large do numbers need to be before you could no longer find factors? 
8. (5) How many codes did you break? Give some examples.
9. (5) Use (by doing an example in the notebook like #4) how using THM 2 page 258 can improve the brute force method breaking RSA code    
10. (10) Conclusion. Return to the questions and comments you asked about RSA in homework 6.  Answer your questions and/or explain what you have learned.

    
Grading - all code must be commented.

**Max grade 60 pts** - Successful implementation of the 1 - 4. Clear and comprehensive write up.

**Max grade 80 pts** - add #5, and #10  additionally include including testing code and exchanging keys at least 3 times. Clear and comprehensive write-up.

To be elegible to get above 80 point **you MUST be exchanging keys by the Friday before the project is due.** You will not get credit for breaking codes on the day it is due as others will not have time to try your codes.

**Max grade 100 pts** - Do all of #1-#10.  Clear and comprehensive write-up.


Do NOT copy code from the internet (or anywhere)
Do not ask questions on stackoverflow or similar etc…
You are expected to have read the honestly notes in the syllabus.

Credit sources you use for additional understanding.
Credit classmates that helped you.